main/src/ca/uwaterloo/flix/language/phase/unification/EffUnification3.scala

```diff
@@ -0,0 +1,308 @@
+          case Status.Timeout(msg) =>
+            Result.Err(UnificationError.TooComplex(msg, eq.loc))
+        }
+
+    }
+  }
+
+  /**
+    * Returns a bi-directional map from type variables to ints computed from the given list of
+    * unification equations `l`.
+    */
+  private def mkBidirectionalVarMap(l: List[(Type, Type, SourceLocation)]): Bimap[Atom, Int] = {
+    // Find all atoms that occur in anywhere in `l`.
+    val atoms = l.foldLeft(Set.empty[Atom]) {
+      case (acc, (t1, t2, _)) => acc ++ getAtoms(t1) ++ getAtoms(t2)
+    }
+
+    // Construct the map from atoms to ints.
+    mkBidirectionalVarMap(atoms)
+  }
+
+  private def mkBidirectionalVarMap(tpe1: Type, tpe2: Type): Bimap[Atom, Int] = {
+    mkBidirectionalVarMap(getAtoms(tpe1) ++ getAtoms(tpe2))
+  }
+
+
+  private def mkBidirectionalVarMap(atoms: Set[Atom]): Bimap[Atom, Int] = {
+    Bimap.from(atoms.toList.zipWithIndex)
+  }
+
+  /** Returns the set of atoms that occur in the given type. */
+  private def getAtoms(t: Type): Set[Atom] = t match {
+    case Type.Var(sym, _) => Set(Atom.Var(sym))
+    case Type.Cst(TypeConstructor.Effect(sym), _) => Set(Atom.Eff(sym))
+    case Type.Cst(TypeConstructor.Error(id, kind), _) => Set(Atom.Error(id, kind))
+    case Type.Cst(_, _) => Set.empty
+    case Type.Apply(tpe1, tpe2, _) => getAtoms(tpe1) ++ getAtoms(tpe2)
+    case Type.Alias(_, _, tpe, _) => getAtoms(tpe)
+    case Type.AssocType(AssocTypeConstructor(sym, _), arg, kind, _) => Set(Atom.Assoc(sym, getAssocArg(arg), kind))
+    case Type.JvmToEff(_, loc) => throw InternalCompilerException("dont know", loc)
+    case Type.JvmToType(_, loc) => throw InternalCompilerException("dont know", loc)
+    case Type.UnresolvedJvmType(_, loc) => throw InternalCompilerException("dont know", loc)
+  }
+
+  /**
+    * Returns a atom corresponding to the type t.
+    *
+    * Fails if t is not a valid argument to a simplified associated typed.
+    */
+  private def getAssocArg(t: Type): Atom = t match {
+    case Type.Var(sym, _) => Atom.Var(sym)
+    case Type.AssocType(AssocTypeConstructor(sym, _), arg, kind, _) => Atom.Assoc(sym, getAssocArg(arg), kind)
+    case Type.Alias(_, _, tpe, _) => getAssocArg(tpe)
+    case tpe => throw InternalCompilerException(s"unexpected associate type argument: $tpe", tpe.loc)
+  }
+
+  /**
+    * Translates the given unification equation on types `p` into a unification equation on terms.
+    */
+  private def toEquation(p: (Type, Type, SourceLocation))(implicit scope: Scope, renv: RigidityEnv, m: Bimap[Atom, Int]): Equation = {
+    val (tpe1, tpe2, loc) = p
+    Equation.mk(toSetFormula(tpe1), toSetFormula(tpe2), loc)
+  }
+
+  /**
+    * Returns the given type `t` as term.
+    *
+    * Uses the given bimap `m` to map type variables to term variables.
+    *
+    * The rigidity environment `renv` is used to map rigid type variables to constants and flexible type variables to term variables.
```

Comment:
How do we treat regions? I would assume we should start treating them as elements-- but maybe that needs a lot more refactor?

Reply:
Yes, thats not this PR. The typing system itself needs to handle concrete regions before this solver can